Initial support for preprocessor procedures

msujew opened the pull request:
Related to https://github.com/zowe/zowe-pli-language-support/issues/297. Adds support for preprocessor procedures in preprocessor code. I.e. you can now call a (locally) defined preprocessor procedure:

```
%TEST_PROC: PROC (...) RETURNS (CHAR);
  // DO SOMETHING
%END;
%VAR = TEST_PROC(...);
```

And expect that the correct value is now attached to `VAR`.

Out of scope for this PR: Use preprocessor procedures outside of preprocessor code. This requires some tricky behavior, due to the `STATEMENT` keyword:

<img width="802" height="394" alt="grafik" src="https://github.com/user-attachments/assets/68218983-a27a-4513-926d-2d59faf6f852" />

msujew committed:
Support preprocessor procedures

Signed-off-by: Mark Sujew <mark.sujew@typefox.io>

ssmifi commented on packages/language/src/preprocessor/pli-preprocessor-parser-state.ts:
```diff
@@ -18,7 +18,10 @@ import { URI } from "../utils/uri";
 
-type ParserLocation = "in-statement" | "in-procedure";
+export enum ParserLocation {
+  Statement,
```
_Question_: Is the `Statement` information in `Location` used anywhere now that `isOnlyInStatement` relies on `Procedure` or will it be needed in the future?

msujew replied:
I've removed it, it was no longer required.

ssmifi commented on packages/language/src/preprocessor/pli-preprocessor-parser.ts:
```diff
@@ -364,6 +349,32 @@ export class PliPreprocessorParser {
     return statement;
```
This `state.pop();` is already handled by the `commonStatement` caller, correct?

_Suggestion_: If so, add a test case that catches this behavior.

msujew replied:
Thanks, fixed this.

ssmifi commented on packages/language/src/preprocessor/pli-preprocessor-parser-state.ts:
```diff
@@ -70,7 +71,7 @@ export class PliPreprocessorParserState implements PreprocessorParserState {
   public uri: URI;
   private text: string;
   private location: ParserLocation[] = [];
```
_Question_: Do you need the stack for nesting? 
According to spec, preprocessor procedures cannot be nested (p. 601, lrm). If this is really the case (and `ParserLocation.Statement` is not required elsewhere), we should prohibit nesting and the `inProcedure` logic could probably be simplified.

msujew replied:
I've removed it, due to the lack of procedure nesting.

ssmifi reviewed (commented):
Code looks good. However, I am unsure about the nesting of preprocessor procedures (see questions.) Although there seems to be an issue with the scoping(?), the parser currently recognizes nested preprocessor procedures, correct? According to spec (p. 601, lrm), this should not be valid.

msujew committed:
Review comments

Signed-off-by: Mark Sujew <mark.sujew@typefox.io>

msujew commented:
@ssmifi Thanks, I didn't look too closely at the list of supported statements in a preprocessor procedure. I've adapted the parser to support the correct ones and removed all code for nested procedures (since the parser rejects those anyway now).

I've also added a few more tests that helped me identify another issue in the parsing code related to `END` statements.

ssmifi commented on packages/language/src/preprocessor/instruction-interpreter.ts:
```diff
@@ -682,29 +771,44 @@ function runDeclareInstruction(
+  // If the local variables are defined, that means we are in a procedure
+  const variables = context.localVariables ?? context.variables;
+  // Don't override existing variables
+  if (!variables.has(instruction.name)) {
```
_Question_: Couldn't there be variables in nested scopes that shadow existing local variables? Is this a thing in the pp?

msujew replied:
Can you give an example? I have a hard time imagining how this would look like.

ssmifi replied:
I was thinking about sth like 
```
 %TEST: PROC (VAR) RETURNS (FIXED);
   DCL X FIXED;
   X = 1;
   DO;
     DCL X FIXED;
     X = 2;
   END;
   RETURN (X);
 %END;
```
but I quickly tried that on the normal parser and it is also not valid, so I guess that's not a thing in pli. Sorry for the delay. :)

msujew replied:
I see. Yes, PLI does not see `DO`, `IF`, etc. as a new scope. Only entering a procedure qualifies as a new scope. We already handle that correctly both in the interpreter and in the validation 👍

ssmifi reviewed (commented):
Thanks, looks good. I have one remaining question about the shadowing of local variables.
